### ISSUE TITLE
Class not assignable to base class type, with method that takes `this` argument

### ISSUE THREAD

**Issue by jcormont** [Microsoft/TypeScript#19661]
**TypeScript Version:**  2.6 (works fine in 2.5)

**Code**

```ts
// a class with 2 methods:
class A {
    f: (t: this) => void;  // declare f
    g(t: this) { /* nop */ }  // define g
}

// a class that adds a single property:
class B extends A { x = 1 }

let a: A = new A();   // OK
let b: B = new B();   // OK

a = b;  // error!?

/*

ERROR:
Type 'B' is not assignable to type 'A'.
  Types of property 'f' are incompatible.
    Type '(t: B) => void' is not assignable to type '(t: A) => void'.
      Types of parameters 't' and 't' are incompatible.
        Type 'A' is not assignable to type 'B'.
          Property 'x' is missing in type 'A'.

(t: B... is not assignable to (t: A...
because A is not assignable to B? <= switched?

*/
```

**Expected behavior:** no errors, since `b` _should_ be assignable to `a`.

**Actual behavior:** error, see above.

_However,_ comment out the declaration for `f` and the error disappears. So, `g` is fine here, even though it is exactly the same type as `f` (for both `a` _and_ `b`).

**Comment by andy-ms** [comment]
Consider this program:
```ts
class A {
    f: (t: this) => void;
}
class B extends A { x = 1 }
class C extends A { y = 1 }

const b: B = new B();
const c: C = new C();

let a: A = b as any;
b.f = (b: B) => {
    console.log(b.x.toPrecision(1));
};
let f: (a: A) => void = a.f;
a = c as any;
a.f = f;

c.f(c); // Oh no! We pass a `c` but at runtime the function is expecting a `b`
```

The problem is that `this` doesn't mean `A`, it means the current *subclass* of `a`.

**Comment by mhegazy** [comment]
consider:

```ts
class B extends A {
    x = 1
    g(t: this) { t.x.toFixed(); }
}

let a: A = new A();   // OK
let b: B = new B();   // OK


a = b; 
a.g(a);  // runtime error
```

**Comment by andy-ms** [comment]
@mhegazy That example compiles!

**Comment by mhegazy** [comment]
> That example compiles!

sorry, should have noted that you need `--strict` to get the error flagged.

**Comment by jcormont** [comment]
@andy-ms Thanks! Now at least I know how to disable this behavior, wasn't aware of `--strictFunctionTypes`. How do I `pragma strictFunctionTypes off` for a specific property...?

To me it's still strange that `g` works while `f` doesn't, what's the difference here? They seem to have the same "signature". If I make `f` readonly, that should make it behave exactly like `g` (can't assign the "wrong" function like in your example), no?

I want to express to the consumer of this class that every sub class of A does indeed accept only its own instances as arguments to `f`. At the level of `A`, it doesn't matter, because all arguments should be of type `A`. At the level of `B`, I need a function type that doesn't allow passing in `C` if that also happens to derive from `A` (unless it derives from `B`, of course).

How would I do that, then? Without defining the actual function... (this is actually an interface in my code, so I can't make it look like `g` here)?

**Comment by jcormont** [comment]
Oh, and of course, I need this `f` function to work, while still being able to assign `Developer` (B) to `Person` (A). Let's say we call the function f something like `collaborateWith(other: this): ...`. You see the use case.

In my case A is also abstract (so like `Person`), which is why a generic `f` makes sense.

Any suggestions?

**Comment by stanhuff** [comment]
@mhegazy @andy-ms In both of your cases, the callback is being called by the client allowing the client to pass an invalid parameter to the function.  When this pattern is being used as an event/signal dispatcher, do you see this same weakness though?  Please grant me some latitude as some of these typing cases can really spin me, but is this code not statically provable to be safe assuming declarations are trustworthy?  If so, is it practical to relax the check to allow this?  Grateful for any insights...

```typescript
class A {
    setChanged(changed: (what: this) => void): void {
        this.changed = changed;
    }
    private changed: (what: this) => void;
    protected fireChanged() {
        if (this.changed)
            this.changed(this);
    }
}

class B extends A {
    private _x: number;
    set x(value: number) {
        if (value != this._x) {
            this._x = value;
            this.fireChanged();
        }
    }
    get x(): number {
        return this._x;
    }
}

declare function fA(a: A) : void;
declare function fB(a: B) : void;

let b = new B();

b.setChanged(fA);   // error - how can this cause an issue if fA is true to its declaration?
b.setChanged(fB);   // no error as expected

let a: A = b;    // error - how can this cause an issue?
```

**Comment by andy-ms** [comment]
@stanhuff I *think* that's sound because `changed` is effectively write-only. If `changed` were readable by clients then it wouldn't be sound: `fA` expects to get an `A` whose `changed` can take an `A`, but if passed a `B` its `changed` will only be able to take a `B` which isn't compatibile with the expected type, which is a callback that can take an `A`. It's likely that we just aren't considering `private` when determining variance.

**Comment by makana** [comment]
I don't think it is working as indented.

Consider the following code:
```typescript
class A
{
	public doSomething(): this
	{
		return this;
	}
}

class B extends A
{

}

class C<T extends A>
{
	public constructor(public readonly o: T)
	{

	}
}

function generic<T extends B>(inst: T): C<T>
{
	return new C<T>(inst); // this fails because...
}
```
...
- "T is not satisfying the constraint A"
  - Type 'B' is not assignable to type 'A'
    - Types of property 'doSomething' are incompatible.
      - Type '() => B' is not assignable to type '() => T'.
        - Type 'B' is not assignable to type 'T'.

It fails even if ``--strictFunctionTypes`` is disabled.

For me, it doesn't matter if B is assignable to T. It is fine that T is assignable to B.

The code works if doSomething() does not use 'this' or if 'generic' uses 'T extends A'

I think the bug is that 'this' is not resolved to type 'T' when using generics and inheritance.

makana

**Comment by andy-ms** [comment]
@makana Moved to #19961
